src/transformers/models/bart/modeling_bart.py

```
@@ -771,6 +754,135 @@ def dummy_inputs(self):
         }
         return dummy_inputs
 
+    def _update_causal_mask(
```

Can we copy this e.g. from llama or whisper?

---

`output_attentions=True & cross_attn_head_mask can not be supported when using SDPA, and we fall back on the manual implementation that requires a 4D causal mask in all cases.`

not really, because of the above comment on orig impl. TL;DR; we need to prepare mask for eager mode if there is an attn-head-mask. All encoder-decoder models therefore have "copied from Bart" now

I feel like the `_update_causal_mask` now has 3 versions: llama, bart and sliding window mixtral. We can try to get one general version copied everywhere in some later PRs

---

The [PR that added the line you mention](https://github.com/huggingface/transformers/pull/26572) also added it to Whisper, and yet when we refactored Whisper we added the copy (from llama) statement on `_update_causal_mask` 🤔 

What breaks if we simply copy this function from llama?

---

hmm afair some tests were failing or I am confusing smth. Let me check what exactly was breaking

---

Found it, `tests/models/bart/test_modeling_bart.py::BartModelTest::test_encoder_decoder_model_standalone`

Oke, so the test can be green by simply passing the `head_mask` so that it is present in both `forward` calls, which later should match outputs logit level. But the main thing to note is: if one passes a `head_mask` to decoder + attention mask with no padding, then one will get no causal mask on decoder part which I believe is not what we want. This will be a breaking change compared to the prev versions

Interestingly same should hold for `layer_head_mask` but it was not added as exception and noone reported issues. So my guess is that using head mask is a rare thing and we can remove this part, taking risk that some user might or might not report it as a bug one day.

I am more for keeping it, just to be correct when creating the mask

---

I see, thank you for digging!

I did some digging on my end too: 
- `head_mask` is a flag that was introduced in 2020 to see whether a given attention head was contributing to the quality of the output. 
- It is no longer present in modern architectures (e.g. llama) and it is not compatible with SDPA nor FA2 (if we check the FA2 code, we see it is not used). 
- There are [no recent issues about it](https://github.com/search?q=repo%3Ahuggingface%2Ftransformers+head_mask&type=issues&s=created&o=desc&p=1).

It seems to me that we are adding/keeping complexity for an old reasearchy feature that is no longer used. Moreover, this issue seems to be present on several models (head mask being ignored in certain models, when SDPA is passed). As such, I believe we should:
1. Simply copy `_update_causal_mask` from llama, to contain complexity and maintenance burden;
2. On all models: `head_mask` + SDPA -> raise exception and ask users to use eager attention

cc @ArthurZucker for input, as this is a decision for multiple architectures :)

---

let's just remove head mask! If people need it they will let us know, but let's just be simple!

---

Removing `head_mask` seems like a good plan to. Keep things simple, let go of old niche features 🙏

---

Opened a new PR for that (https://github.com/huggingface/transformers/pull/35786). Will come back to Bart compile after it is merged